chia/types/borderlands.py

```diff
@@ -0,0 +1,28 @@
+from __future__ import annotations
+
+from typing import NewType
+
+from chia.types.blockchain_format.sized_bytes import bytes48
+
+PublicKeyBytes = NewType("PublicKeyBytes", bytes48)
+
+
+"""
+These conversion functions should be used at the border between native code (eg. Rust and C++) and Python
+
+Use bytes_to_TypeName() for bytes coming from native code into the Python program.
+Use the TypeName() constructor if you are constructing e.g. a bytes32 from within Python, perhaps for use in a test.
+
+Note the type signature of the conversion functions:
+```
+    def bytes_to_PublicKeyBytes(input_bytes: bytes) -> PublicKeyBytes:
+```
+
+Native modules are only able to return native Python types by default, thus we get `bytes`, and not `bytes32`
+from those libraries.
+
+"""
+
+
+def bytes_to_PublicKeyBytes(input_bytes: bytes) -> PublicKeyBytes:
+    return PublicKeyBytes(bytes48(input_bytes))
```

Comment:
True. To change this, we can add a runtime check like this

```
def bytes_to_PublicKeyBytes(input_bytes: bytes) -> PublicKeyBytes:
    return PublicKeyBytes(bytes48(input_bytes))
```

to get the minor benefit of encouraging conversions closer to the source, or we can keep a more primitive type conversion, keeping the TypeName constructors, and tossing the bytes_to_TypeName conversions.

@altendky @arvidn Thoughts?

Reply:
Also, if we want the types converted closer then maybe that should be done in the library?  I've suggested a few times that it would be nice for all our projects to have the outer layer be a Python package.  In this case that would let us add wrapper functions that could do this hinting/runtime conversion and checks for us.  For things that don't need wrapping, they can just be re-exported to avoid any runtime overhead or the need to write wrapper methods.

Another direction, make our fixed width bytes and integer types just `NewType`s as well.  Then, at the boundaries of serialization (network, db, etc) we can deal with the serialization.  This has some complexity around checks and is not particularly tied to this change here.  Just commenting in case discussion would be of interest.

Reply:
> Also, if we want the types converted closer then maybe that should be done in the library? I've suggested a few times that it would be nice for all our projects to have the outer layer be a Python package.

I think everyone would prefer that, but it takes work, and it would have to be higher priority than other things.

Reply:
> It's still not clear to me when bytes_to_PublicKeyBytes() will do something useful. Both for hinting and at runtime it will accept bytes, bytes48, and PublicKeyBytes as I understand it. At runtime, PublicKeyBytes will do what bytes48 does so bytes_to_PublicKeyBytes(x) will be the same as bytes48(bytes48(x)) I think.

No, I don't think you are correct.

```
$ cat x.py
from typing import NewType
from chia.types.blockchain_format.sized_bytes import bytes32


b = b"b"
b32 = bytes32(b"b" * 32)

T = NewType("T", bytes32)

a = T(b)
b = T(b32)

$ mypy x.py
x.py:10: error: Argument 1 to "T" has incompatible type "bytes"; expected "bytes32"  [arg-type]
Found 1 error in 1 file (checked 1 source file)

```

Reply:
I think you are right and have now gotten me caught up.  I think I was off on my thinking about inheritance and also the direction I was looking at this from.  Thanks.

Now I'm wondering if it would be worth avoiding the function call overheads for the hint-only benefit.  It both seems like over-optimization and also appropriate given that we are optimizing this to the point of writing the implementations in rust.  `:]`


```python
from typing import TypeVar
from typing_extensions import Protocol

_T = TypeVar("_T", covariant=True)

class BytesToProtocol(Protocol[_T]):
    def __call__(self, __positional_only: bytes) -> _T:
        ...

bytes_to_PublicKeyBytes: BytesToProtocol[PublicKeyBytes] = bytes48
```

Please do question the appropriateness of covariance there.  mypy wanted it and I have to refresh myself every time I deal with this.

These examples show what works and what complains and needs an ignore (for the example, in reality this is what you intend to not allow, you want the mypy complaint)

```python
x: PublicKeyBytes

x = bytes_to_PublicKeyBytes(b"")
x = bytes_to_PublicKeyBytes(bytes48(b""))
x = bytes_to_PublicKeyBytes(PublicKeyBytes(bytes48(b"")))

x = PublicKeyBytes(b"")  # type: ignore[arg-type]
x = PublicKeyBytes(bytes48(b""))
x = PublicKeyBytes(PublicKeyBytes(bytes48(b"")))
```

Perhaps we don't even care about the actual object type and would be happy to not even convert it and trust that the library gave us the right thing.  That's what we have now in at least some cases I think where rust claims `bytes32` but really has only created `bytes`.  Am I remembering correctly @arvidn?  If we're ok with that, then instead of `= bytes48` it could be `= lambda __x: __x`.  (I don't know of an optimized no-op that can be written as a function call)

I have no evidence that these are faster.  Just "reasoning", which isn't always right.  And even if it is, the amount may not be of interest.

Reply:
I'm totally on board with the core concept of these PRs.  That's why I care to talk through the implementation.  `:]`

I don't think rust, as we handle it now, can literally give us `bytes32` etc.  If nothing else, that'd make a circular type dependency.  Though I think Arvid has commented more about it being an aspect of maturin.  But...  I know I was displeased about the hints lying about the type somewhere.  Looking elsewhere in the same `.pyi` you find things like `def tree_hash(program: ReadableBuffer) -> bytes32: ...`.  To be clear, displeased means I hope we can find something better, not that I know this was the wrong choice.

So, in those cases we ought to be able to just update the existing hint generation to reference these new types instead?

Reply:
There are two main challenges in making rust code return the "rich" python types like `bytes32` or `PublicKeyBytes`:

1. Since those types are defined in `chia-blockchain`, it would create a circular dependency
2. It's not trivial to construct native python types and return on the rust side

Those challenges aren't impossible, and one way to solve them (although, perhaps, heavy handed) would be to move `Bytes32` et.al. into `chia_rs` and make them rust native.

If it would help, we could define these python types in `chia_rs.pyi`. It might still be problematic constructing actual plain `bytes` objects and hint them to be `bytes32` or `PublicKeyBytes`

Reply:
I think it's reasonable to have to cast at the python/rust boundary for now, in the interest of making progress. I'm not so sure about casting in other places though.

Reply:
By cast, do you mean a hint-only thing?  Or actually creating the `bytes32` etc objects.  In at least some cases now we are avoiding the object creation overhead and just doing a hint-only 'cheat' in the `.pyi`.  If we leave that and then _add_ any call in chia-blockchain we are adding new overhead.  Even if just `lambda _x: _x`.  This may not be critical.

Reply:
I think the best next-step following the theme of these patches is to move bytes32, etc. over to `clvm_tools_rs`

It will be a big pain initially, but once they are over there, it will be easy to export types like these that build on them.

As for these patches, though, if we keep the changes to the Python side only, I think the right thing to do is do any checks (both static and dynamic) once at the boundary, and just change the type annotations everywhere else.

 In its current state, this patch only does static checks at the interface, there are no dynamic checks, and there are no conversions or object creations afterward. You can check that for yourself by searching for `PublicKeyBytes(`

Reply:
Well, I apparently entirely misunderstood the runtime behavior of `PublicKeyBytes()`.  `:[`  I thought it acted like the constructor of the type it was "created from" as opposed to being a runtime no-op.  I'm sorry that I missed that.  It's right there...

https://docs.python.org/3.11/library/typing.html#newtype
> At runtime, the statement `Derived = NewType('Derived', Base)` will make `Derived` a callable that immediately returns whatever parameter you pass it.

I'm good with resolving this discussion.  I'll add a thumbs up here and once you to do as well then whoever can mark it resolves.

I think I would need to hear more about `bytes32` etc going into `clvm_tools_rs`, but that's another discussion we can take up elsewhere.  Thanks for talking through this all with me.